File: app/src/main/java/com/recursia/popularmovies/di/modules/MapperModule.java
package com.recursia.popularmovies.di.modules;

import com.recursia.popularmovies.data.db.mappers.CastDatabaseModelToEntityMapper;
import com.recursia.popularmovies.data.db.mappers.EntityToCastDatabaseModelMapper;
import com.recursia.popularmovies.data.db.mappers.EntityToGenreDatabaseModelMapper;
import com.recursia.popularmovies.data.db.mappers.EntityToMovieExtraDatabaseModelMapper;
import com.recursia.popularmovies.data.db.mappers.EntityToTrailerDatabaseModelMapper;
import com.recursia.popularmovies.data.db.mappers.GenreDatabaseModelToEntityMapper;
import com.recursia.popularmovies.data.db.mappers.MovieDatabaseModelToEntityMapper;
import com.recursia.popularmovies.data.db.mappers.MovieExtraDatabaseModelToEntityMapper;
import com.recursia.popularmovies.data.db.mappers.TrailerDatabaseModelToEntityMapper;
import com.recursia.popularmovies.data.network.mappers.CreditsResponseToCastMapper;
import com.recursia.popularmovies.data.network.mappers.DiscoverMovieResponseToMovieMapper;
import com.recursia.popularmovies.data.network.mappers.MovieTrailersResponseToTrailersMapper;

import dagger.Module;
import dagger.Provides;

@Module
public class MapperModule {
    @Provides
    CastDatabaseModelToEntityMapper provideCastDatabaseModelToEntityMapper() {
        return new CastDatabaseModelToEntityMapper();
    }

    @Provides
    CreditsResponseToCastMapper provideCreditsResponseToCaseMapper(CastDatabaseModelToEntityMapper mapper) {
        return new CreditsResponseToCastMapper(mapper);
    }

    @Provides
    DiscoverMovieResponseToMovieMapper provideDiscoverMovieResponseToMovieMapper(MovieDatabaseModelToEntityMapper mapper) {
        return new DiscoverMovieResponseToMovieMapper(mapper);
    }

    @Provides
    GenreDatabaseModelToEntityMapper provideGenreDatabaseModelToEntityMapper() {
        return new GenreDatabaseModelToEntityMapper();
    }

    @Provides
    MovieDatabaseModelToEntityMapper provideMovieDatabaseModelToEntityMapper(GenreDatabaseModelToEntityMapper mapper) {
        return new MovieDatabaseModelToEntityMapper(mapper);
    }

    @Provides
    EntityToMovieExtraDatabaseModelMapper provideEntityToMovieDatabaseModelMapper(EntityToCastDatabaseModelMapper entityToCastDatabaseModelMapper,
                                                                                  EntityToGenreDatabaseModelMapper entityToGenreDatabaseModelMapper,
                                                                                  EntityToTrailerDatabaseModelMapper entityToTrailerDatabaseModelMapper) {
        return new EntityToMovieExtraDatabaseModelMapper(entityToCastDatabaseModelMapper, entityToGenreDatabaseModelMapper, entityToTrailerDatabaseModelMapper);
    }

    @Provides
    EntityToTrailerDatabaseModelMapper provideEntityToTrailerDatabaseModelMapper() {
        return new EntityToTrailerDatabaseModelMapper();
    }

    @Provides
    EntityToGenreDatabaseModelMapper provideEntityToGenreDatabaseModelMapper() {
        return new EntityToGenreDatabaseModelMapper();
    }

    @Provides
    EntityToCastDatabaseModelMapper provideEntityToCastDatabaseModelMapper() {
        return new EntityToCastDatabaseModelMapper();
    }

    @Provides
    MovieTrailersResponseToTrailersMapper provideMovieTrailersResponseToTrailersMapper(TrailerDatabaseModelToEntityMapper mapper) {
        return new MovieTrailersResponseToTrailersMapper(mapper);
    }

    @Provides
    TrailerDatabaseModelToEntityMapper provideTrailerDatabaseModelToEntityMapper() {
        return new TrailerDatabaseModelToEntityMapper();
    }

    @Provides
    MovieExtraDatabaseModelToEntityMapper provideMovieExtraDatabaseModelToEntityMapper(GenreDatabaseModelToEntityMapper genreDatabaseModelToEntityMapper,
                                                                                       CastDatabaseModelToEntityMapper castDatabaseModelToEntityMapper,
                                                                                       TrailerDatabaseModelToEntityMapper trailerDatabaseModelToEntityMapper) {
        return new MovieExtraDatabaseModelToEntityMapper(genreDatabaseModelToEntityMapper, castDatabaseModelToEntityMapper, trailerDatabaseModelToEntityMapper);
    }
}

File: app/src/main/java/com/recursia/popularmovies/di/modules/MapperModule.kt
package com.recursia.popularmovies.di.modules

import com.recursia.popularmovies.data.mappers.*
import dagger.Module
import dagger.Provides

@Module
class MapperModule {

    @Provides
    internal fun provideCastDatabaseModelToEntityMapper(): CastModelToEntityMapper {
        return CastModelToEntityMapper()
    }

    @Provides
    internal fun provideCreditsResponseToCaseMapper(mapper: CastModelToEntityMapper): CreditsResponseToCastMapper {
        return CreditsResponseToCastMapper(mapper)
    }

    @Provides
    internal fun provideDiscoverMovieResponseToMovieMapper(mapper: MovieModelToEntityMapper): DiscoverMovieResponseToMovieMapper {
        return DiscoverMovieResponseToMovieMapper(mapper)
    }

    @Provides
    internal fun provideGenreDatabaseModelToEntityMapper(): GenreModelToEntityMapper {
        return GenreModelToEntityMapper()
    }

    @Provides
    internal fun provideMovieDatabaseModelToEntityMapper(mapper: GenreModelToEntityMapper): MovieModelToEntityMapper {
        return MovieModelToEntityMapper(mapper)
    }

    @Provides
    internal fun provideEntityToTrailerDatabaseModelMapper(): EntityToTrailerModelMapper {
        return EntityToTrailerModelMapper()
    }

    @Provides
    internal fun provideEntityToGenreDatabaseModelMapper(): EntityToGenreModelMapper {
        return EntityToGenreModelMapper()
    }

    @Provides
    internal fun provideEntityToCastDatabaseModelMapper(): EntityToCastModelMapper {
        return EntityToCastModelMapper()
    }

    @Provides
    internal fun provideMovieTrailersResponseToTrailersMapper(mapper: TrailerModelToEntityMapper): MovieTrailersResponseToTrailersMapper {
        return MovieTrailersResponseToTrailersMapper(mapper)
    }

    @Provides
    internal fun provideTrailerDatabaseModelToEntityMapper(): TrailerModelToEntityMapper {
        return TrailerModelToEntityMapper()
    }

    @Provides
    internal fun provideReviewNetworkModelToEntityMapper(): ReviewNetworkToEntityModelMapper {
        return ReviewNetworkToEntityModelMapper()
    }

    @Provides
    internal fun provideReviewsResponseToReviewMapper(
        reviewNetworkToEntityModelMapper: ReviewNetworkToEntityModelMapper
    ): ReviewsResponseToReviewMapper {
        return ReviewsResponseToReviewMapper(reviewNetworkToEntityModelMapper)
    }

    @Provides
    internal fun provideGenresResponseToEntityMapper(
        genreModelToEntityMapper: GenreModelToEntityMapper
    ): GenresResponseToEntityMapper {
        return GenresResponseToEntityMapper(genreModelToEntityMapper)
    }
}
